Fixed #24119, #24120 -- Formalized debug integration for template backends

prestontimmons opened the pull request.

prestontimmons commented:
High-level overview of changes:

Commit 1:

* Added backend and chain arguments to TemplateDoesNotExist.
* Updated ``django.template.backends`` to reraise TemplateDoesNotExist with backend specified.
* Updated ``django.template.loader`` methods to raise TemplateDoesNotExist with chain specified.
* Removed old logic from debug view.
* Integrated Jinja2 traceback information into debug page.

Commit 2:

* Moved TemplateDoesNotExist and TemplateSyntaxError to django.template.exceptions since these are no longer DTL-specific.
* Normalized import usage:
  * modules in ``django.template`` use relative imports
  * modules outside use ``django.template`` rather than ``django.template.base``

timgraham commented on django/template/backends/dummy.py:
```diff
@@ -29,17 +30,26 @@ def from_string(self, template_code):
 
         else:
-            raise TemplateDoesNotExist(template_name)
+            raise TemplateDoesNotExist(
```
I'd opt for 1 line

timgraham commented on django/template/backends/jinja2.py:
```diff
@@ -61,3 +65,29 @@ def render(self, context=None, request=None):
+    """
+    Formats exception information for display on the debug page.
+    """
+    context_lines = 10
```
I guess the name/purpose of these variables are documented elsewhere, but maybe we could at least add a pointer to that. They aren't self-explanatory.

prestontimmons replied:
Okay. Do you think it would be clearer if this returned a class instance, instead of a dict?

```
return TemplateExceptionDetail(
    name=...
   message=...
)
```

Then there would be an obvious place to document it.

timgraham replied:
No strong preference from me.

timgraham commented on django/template/exceptions.py:
```diff
@@ -0,0 +1,31 @@
+class TemplateDoesNotExist(Exception):
+    """
+    The central exception used by backends when a template does not exist.
```
not sure exactly what's meant by "central exception"

timgraham commented on docs/topics/templates.txt:
```diff
@@ -149,14 +149,35 @@ The ``django.template.loader`` module defines two functions to load templates.
 .. currentmodule:: django.template
 
-If loading a template fails, the following two exceptions, defined in
+If loading a template fails the following two exceptions, defined in
```
the removed comma seems fine here

timgraham commented:
Doc edits below. Could you also add a mention in the 1.9 release notes?
```diff
diff --git a/docs/topics/templates.txt b/docs/topics/templates.txt
index 7516717..da580b0 100644
--- a/docs/topics/templates.txt
+++ b/docs/topics/templates.txt
@@ -149,23 +149,23 @@ The ``django.template.loader`` module defines two functions to load templates.
 
 .. currentmodule:: django.template
 
-If loading a template fails the following two exceptions, defined in
+If loading a template fails, the following two exceptions, defined in
 ``django.template``, may be raised:
 
 .. exception:: TemplateDoesNotExist(msg, tried=None, backend=None, chain=None)
 
     This exception is raised when a template cannot be found. It accepts the
-    following optional arguments for populating the
-    :ref:`template postmortem <template-postmortem>` on the debug page:
+    following optional arguments for populating the :ref:`template postmortem
+    <template-postmortem>` on the debug page:
 
     ``backend``
-        The template backend instance the exception originated from.
+        The template backend instance from which the exception originated.
 
     ``tried``
-        A list of sources that were tried when finding the template. This
-        is formatted as a list of tuples containing (origin, status), where
-        origin is an :class:`~django.template.base.Origin` object and status
-        is a string with the reason the template wasn't found.
+        A list of sources that were tried when finding the template. This is
+        formatted as a list of tuples containing (origin, status), where origin
+        is an :class:`~django.template.base.Origin` object and status is a
+        string with the reason the template wasn't found.
 
     ``chain``
         A list of intermediate :exc:`~django.template.TemplateDoesNotExist`
@@ -569,15 +569,15 @@ Custom engines can populate this information by setting a ``template_debug``
 attribute on exceptions raised during parsing and rendering. This attribute
 is a :class:`dict` with the following values:
 
-* ``'name'``: The name of the template the exception occurred in.
+* ``'name'``: The name of the template in which the exception occurred.
 
 * ``'message'``: The exception message.
 
 * ``'source_lines'``: The lines before, after, and including the line the
   exception occurred on. This is for context, so it shouldn't contain more than
-  20 lines or so total.
+  20 lines or so.
 
-* ``'line'``: The line number the exception occurred on.
+* ``'line'``: The line number on which the exception occurred.
 
 * ``'before'``: The content on the error line before the token that raised the
   error.
@@ -598,7 +598,16 @@ Given the above template error, ``template_debug`` would look like::
     {
         'name': '/path/to/template.html',
         'message': "Invalid block tag: 'syntax'",
-        'source_lines': [(1, 'some\n'), (2, 'lines\n'), (3, 'before\n'), (4, 'Hello {% syntax error %} {{ world }}\n'), (5, 'some\n'), (6, 'lines\n'), (7, 'after\n'), (8, '')],
+        'source_lines': [
+            (1, 'some\n'),
+            (2, 'lines\n'),
+            (3, 'before\n'),
+            (4, 'Hello {% syntax error %} {{ world }}\n'),
+            (5, 'some\n'),
+            (6, 'lines\n'),
+            (7, 'after\n'),
+            (8, ''),
+        ],
         'line': 4,
         'before': 'Hello ',
         'during': '{% syntax error %}',
```

aaugustin commented on django/template/backends/jinja2.py:
```diff
@@ -61,3 +65,29 @@ def render(self, context=None, request=None):
+    top = max(0, lineno - context_lines - 1)
+    bottom = min(total, lineno + context_lines)
+
+    return dict(
```
The literal syntax is generally more readable.

aaugustin commented on django/template/exceptions.py:
```diff
@@ -0,0 +1,31 @@
+    """
+    def __init__(self, msg, tried=None, backend=None, chain=None):
+        self.backend = backend
+        self.tried = tried or []
```
Can you check explicitly?

    if tried is None:
        tried = []
    self.tried = tried

This is a good practice in general to avoid hard-to-diagnose errors if an unexpected, falsy value of `tried` is passed in.

aaugustin commented:
I read the diff carefully.

Passing a reference to the backend to the template makes for some weird code, but I have convinced myself that it is needed.

prestontimmons commented:
Pushed an update. This comes with some additional changes:

* Added "Origin API and 3rd-party integration" to the docs
* Changed wording where "origin" is referenced since this is more API via convention than enforcement
* Added an Origin object to the jinja2 backend

timgraham commented on django/template/exceptions.py:
```diff
@@ -0,0 +1,43 @@
+"""
+This module contains generic exceptions used by template backends. Although,
+due to historical reasons, the Django template language also internally uses
+these exceptions, other exceptions specific to only DTL should not be added
```
only -> the

timgraham commented on docs/topics/templates.txt:
```diff
@@ -510,6 +530,116 @@ fictional ``foobar`` template library::
+
+.. versionadded:: 1.9
+
+    Debug page integration for non-Django template engines was added in Django
```
since this renders as: "New in Django 1.9: ..." the "in Django 1.9" bit is redundant.

timgraham commented on docs/topics/templates.txt:
```diff
@@ -510,6 +530,116 @@ fictional ``foobar`` template library::
+
+Django templates have an :class:`~django.template.base.Origin` object available
+through the ``template.origin`` attribute. This enables debug information to be
+displayed in the :ref:`template postmortem <template-postmortem>`, as well as
```
as well as in

prestontimmons commented:
This is merged. Thanks for the review!
